### DISCUSSION TITLE
Message queue alignment/size requirements

### DISCUSSION THREAD

**Discussion by devanlai** [zephyrproject-rtos/zephyr#38468] [General]
I posted this on Slack a while ago, but I figured this discussion would be better somewhere a little more accessible.

I've been comparing the [Message Queue](https://docs.zephyrproject.org/latest/reference/kernel/data_passing/message_queues.html) documentation to the implementation and some of the usage prescriptions seem a bit off.

In particular, the documentation for the message queue states:
> The message queue’s ring buffer must be aligned to an N-byte boundary, where N is a power of 2 (i.e. 1, 2, 4, 8, …). To ensure that the messages stored in the ring buffer are similarly aligned to this boundary, the data item size must also be a multiple of N.

And the [Data Passing table](https://docs.zephyrproject.org/latest/reference/kernel/index.html#data-passing) on the Kernel Services documentation page summarizes:
> | Object        | Bidirectional? | Data structure | Data item size | Data Alignment | ISRs can receive? | ISRs can send? | Overrun handling             |
> | ------------- | -------------- | -------------- | -------------- | -------------- | ----------------- | -------------- | ---------------------------- |
> | Message queue | No             | Ring buffer    | Power of two   | Power of two   | Yes [3]           | Yes            | Pend thread or return -errno |

Discounting the liberal interpretation of 2^0 = 1, which would effectively be the same as not having any alignment or size restrictions, it seems to me like these restrictions don't quite make sense given the implementation.

Looking at the source implementation in [msg_q.c](https://github.com/zephyrproject-rtos/zephyr/blob/main/kernel/msg_q.c), all access to the ring buffer contents is done through `memcpy()`. My understanding of `memcpy()` is that it should work correctly regardless of the alignment of the source or destination, though it may be slower when operating on a source or destination that are not aligned to the CPU's native word size.

Based on this, I think it doesn't really make sense to frame the alignment requirements as being related to a power-of-two at all. Instead, I think that the size and alignment can be documented as "arbitrary" with a footnote that performance may be degraded if the alignment and size are not a multiple of the native word size.

**Comment by yashi** [comment]
I've done a bit of history digging:
- The current `memcpy()` implementation came from https://github.com/zephyrproject-rtos/zephyr/commit/456c6daa9f0c706131cff4eedcf3290fdfcf98e9
- The current alignment note in the doc came from https://github.com/zephyrproject-rtos/zephyr/commit/da827225341c35735f00b10db57db7f4ad2e1347
- Both changes are within v1.5 -> v1.6

In the commit for the doc change, even thought it's not for the message queue but memory slab, we find the string "ZEP-1265". And thankfully it's imported to the github as the issue #2747.  The issue says that it was Arc which failed with unaligned access.

I agree that `memcpy()` should be capable of accessing any address. And the `memcpy()` in the minimal libc seems to handle non-aligned access.  So, maybe ring buffer restriction and the ZEP-1265 is unrelated.

Looking forward to someone who knows the history to jump in.
